doc/sphinx/addendum/universe-polymorphism.rst

```diff
@@ -977,6 +977,233 @@ Similar to universes, fresh global sorts can be declared with the :cmd:`Sort`.
+  not appear with an increment (i.e. if `u` is a template level it
+  does not appear as `u+1`).
+
+  This restriction prevents generating increments higher than 1 by
+  applying template inductives to themselves, i.e. if `I : Type@{u} ->
+  Type@{u + 1}` was template polymorphic and `X:Type@{i}` then `I
+  (I X) : Type@{i + 2}`. It is necessary as the current universe
+  checking cannot handle such increments properly, but this
+  restriction may be removed in the future.
+
+- the template sorts appear in the conclusions of parameters (not
+  necessarily linearly), may appear in the inductive's conclusion, and
+  appear nowhere else.
+
+- for each template universe `u`, there are no constraints of the form
+  `v <= u` ("no constraints from below").
+
+These requirements are more than sufficient to ensure that if the
+inductive was declared universe polymorphic and cumulative then all
+template levels would be irrelevant. The "no constraints from below"
+requirement is needed for subject reduction (preservation of typing by
+reduction) in terms involving partially applied template polymorphic
+inductives. Linearity of template universes and not supporting mutual
+inductives make implementation of template polymorphism easier.
+
+.. warning::
+
+   The restriction that universes are introduced by the inductive
+   declaration prevents inductive types declared in sections from being
+   template-polymorphic on universes introduced previously in the
+   section: they cannot parameterize over the universes introduced with
+   section variables that become parameters at section closing time, as
+   these may be shared with other definitions from the same section
+   which can impose constraints on them.
+
+.. note::
+
+   Currently user syntax does not support explicit sort polymorphism
+   annotations (`@{s;...}`) in template polymorphic declarations. The
+   sorts must be left implicit and will be automatically inferred.
+
+.. example::
+
+   .. rocqtop:: in reset
+
+      Inductive option (A:Type) : Type :=
+      | None : option A
+      | Some : A -> option A.
+
+      Inductive prod A B := pair : A -> B -> prod A B.
+
+   .. rocqtop:: all
+
+      Set Printing Universes.
+      About option.
+      About prod.
+
+   Since the sort polymorphism of template inductives cannot be
+   specified in the user syntax, it is instead described by the
+   "can/cannot be instantiated to `Prop`" mention in :cmd:`About`'s
+   output. Since `option` has 2 constructors, it cannot be
+   instantiated to `Prop` i.e. it is not sort polymorphic.
+
+Using template polymorphic inductives
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For each template universe `u` bound by parameter `A` of an inductive
+`I`, when `I` is applied such that `A` is instantiated with a term of
+type `forall ..., Type@{q;i}` (where `i` may be an algebraic
+universe), `u` is instantiated by `i` in the inductive's conclusion
+and in the constraints (which must be of the form `x <= u` where `x`
```

Comment:
That's backwards I think. There can only be `u <= x` constraints so that `max(i0...im) <= x` can be implemented currently.